packages/protocol/contracts/dapis/rrpDapiServer.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.4;
+
+import "../AirnodeRrp.sol";
+import "./CustomReducer.sol";
+
+contract RrpDapiServer is CustomReducer {
+    struct DapiParameters {
+        uint256 minResponsesToReduce;
+        int256 toleranceInPercentages;
+        uint256 requesterIndex;
+        bytes32[] templateIds;
+        address[] designatedWallets;
+        address reduceAddress;
+        bytes4 reduceFunctionId;
+        }
+    
+    AirnodeRrp public airnodeRrp;
+    mapping(uint256 => DapiParameters) private dapiIdToParameters;
+    uint256 public nextDapiID = 1;
+    uint256 public nextDapiRequestIndex = 1;
+    mapping(bytes32 => uint256) private requestIdToDapiRequestIndex;
+    mapping(uint256 => int256[]) private dapiRequestIndexToResponses;
+    mapping(uint256 => uint256) private dapiRequestIndexToDapiId;
+
+    constructor(address _airnodeRrp) {
+        airnodeRrp = AirnodeRrp(_airnodeRrp);
+    }
+
+    function registerDapi
+    (
+        uint256 minResponsesToReduce,
+        int256 toleranceInPercentages,
+        uint256 requesterIndex,
+        bytes32[] calldata templateIds,
+        address[] calldata designatedWallets,
+        address reduceAddress,
+        bytes4 reduceFunctionId
+    )
+        external
+        returns (uint256 dapiId)
+    {
+      dapiIdToParameters[nextDapiID] = DapiParameters(
```

Comment:
I wonder how expensive it would be to compute `dapiId` by taking a has of its parameters. Otherwise, you will have to make a transaction to register a dAPI and then wait enough confirmations to know that your dAPI index will not change.

Reply:
I thought it would be too expensive without any upside, that's why I changed it.

> to know that your dAPI index will not change

Why would it change?

Reply:
The transactions may be reordered if multiple dAPIs are being created in a short period. I'm not sure how much more expensive it would be, probably not that much.